MyRangesImplementation/MyRangesImplementation/filter_range_adapter.hpp

```diff
@@ -0,0 +1,91 @@
+#pragma once
+#include"range_traits.hpp"
+#include"filter_iterator.hpp"
+#include<type_traits>
+#include<vector>
+
+namespace ranges {
+	namespace view {
+		template<typename Range, typename FilterFunc>
+		class filter_range_adapter {
+			static_assert(RangeTraits::isRange<Range>(), "error: the given first type must be a Range type");
+			const Range& m_range;
+			FilterFunc m_filter;
+
+		public:
+			using value_type = typename Range::value_type;
+			using iterator = ranges::internals::iterators::filter_iterator<typename Range::iterator, FilterFunc>;
+
+			filter_range_adapter(const Range& range, FilterFunc filter)
+				noexcept(std::is_move_constructible_v<FilterFunc>)
+				:m_range(range), m_filter(std::move(filter)) {}
+
+			filter_range_adapter(const filter_range_adapter<Range, FilterFunc>& other)
+				noexcept(std::is_copy_constructible_v<FilterFunc>)
+				:m_range(other.m_range), m_filter(other.m_filter) {}
+
+			filter_range_adapter(filter_range_adapter<Range, FilterFunc>&& other)
+				noexcept(std::is_move_constructible_v<FilterFunc>)
+				:m_range(other.m_range), m_filter(std::move(other.m_filter)) {}
+
+			auto begin() const {
+				return ranges::internals::iterators::filter_iterator(this->m_range.begin(), this->m_range.end(), this->m_filter);
+			}
+
+			auto end() const {
+				return ranges::internals::iterators::filter_iterator(this->m_range.end(), this->m_range.end(), this->m_filter);
+			}
+
+			decltype(auto) toVector() const {
+				std::vector<value_type> result;
+				std::copy(this->begin(), this->end(), std::back_inserter(result));
+				return result;
+			}
+
+			operator std::vector<value_type>() const {
+				return this->toVector();
+			}
+		};
+	}
+}
+
+namespace ranges {
+	namespace internals {
+		namespace adaptorFactories {
+			template<typename FilterFunc>
+			class filter_range_adaptor_factory {
```

Comment:
why a class and not a make_filter_range_adapter function?

Reply:
there is a filter function that gets a filter function and return factory that can get the underlying range using operator | and create full filter_range_adapter. To support the syntax of operator | the filter function must return factory (at least in my design)